inc/hclib-async.h

```diff
@@ -55,79 +55,30 @@ namespace hclib {
  * TODO optimize that overhead.
  */
 
-/*
- * At the lowest layer in the call stack before entering user code, this method
- * invokes the user-provided lambda.
- */
-template <typename T>
-inline void call_lambda(T* lambda) {
-	const int wid = current_ws()->id;
-	MARK_BUSY(wid);
-	(*lambda)();
-	HC_FREE((void*) lambda);
-	MARK_OVH(wid);
-}
-
-/*
- * Store a reference to the type-specific function for calling the user lambda,
- * as well as a pointer to the lambda's location on the heap (through which we
- * can invoke it). async_arguments is stored as the args field in the task_t
- * object for a task, and passed to lambda_wrapper.
- */
-template <typename Function, typename T1>
-struct async_arguments {
-    Function lambda_caller;
-    T1 lambda_on_heap;
-
-    async_arguments(Function k, T1 a) :
-        lambda_caller(k), lambda_on_heap(a) { }
-};
-
-/*
- * The method called directly from the HC runtime, passed a pointer to an
- * async_arguments object. It then uses these async_arguments to call
- * call_lambda, passing the user-provided lambda.
- */
-template<typename Function, typename T1>
-void lambda_wrapper(void *args) {
-    async_arguments<Function, T1> *a =
-        (async_arguments<Function, T1> *)args;
-
-    (*a->lambda_caller)(a->lambda_on_heap);
-}
-
-/*
- * Initialize a task_t for the C++ APIs, using a user-provided lambda.
- */
-template<typename Function, typename T1>
-inline void initialize_task(hclib_task_t *t, Function lambda_caller,
-        T1 *lambda_on_heap) {
-    async_arguments<Function, T1 *> *args =
-        new async_arguments<Function, T1*>(lambda_caller, lambda_on_heap);
-    t->_fp = lambda_wrapper<Function, T1 *>;
-    t->args = args;
-    t->is_async_any_type = 0;
-    t->future_list = NULL;
-    t->place = NULL;
+template<typename T>
+void lambda_wrapper(void *arg) {
+    T* lambda = static_cast<T*>(arg);
+    MARK_BUSY(current_ws()->id);
+    (*lambda)(); // !!! May cause a worker-swap !!!
```

Comment:
There are similar comments a few places in the C code. This a reminder that you can't reuse worker-specific values after this function call—e.g., `current_ws()->id`—since the task might block inside the user code, resulting in a context switch.